### PR TITLE
stat_sinks: add OpenTelemetry stats sink

### LINKED ISSUE #26498: Open Telemetry stats sink support
I can see that [stats sinks ](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/metrics/v3/stats.proto#config-metrics-v3-statssink) currently do not have an open telemetry extension. Is there currently WIP to support it or are there any known issues with supporting OTEL metrics sink? If not, will appreciate marking this issue as help wanted.

Implementation-wise, it seems like "Metrics service" extension is doing most of the work, but it's using Prometheus metrics. Maybe it's possibly that most of the code can be extracted to common folder and have "Metrics service" and "otel_grpc" extension share the common code?

cc @snowp

### PR DESCRIPTION
**Pull request by ohadvano**
Commit Message: stat_sinks: OpenTelemetry stats sink
Additional Description: Adding OpenTelemetry stats sink, following the [OTLP](https://opentelemetry.io/docs/reference/specification/protocol/otlp) protocol and [proto definitions](https://github.com/open-telemetry/opentelemetry-proto/tree/main/opentelemetry/proto).
Risk Level: Low
Testing: Unit tests, Integration tests
Docs Changes: Changelog, Stats sinks docs
Release Notes: None
Platform Specific Features: None

Resolves #26498

### CODE REVIEW

**Comment by ohadvano** [comment]
I am opening this PR for initial review, before proceeding to tests implementation. Will appreciate your help with initial review. 
/assign-from @envoyproxy/senior-maintainers
@mattklein123 I have temporarily mentioned you in the CODEOWNERS for that new extension, just to unblock the CI (or not temporarily if requested of course).
@RyanTheOptimist For some reason, the CI passes but when building locally, I get the following error:
> Executing genrule @io_opentracing_cpp//:generate_version_h failed

Seems like this had happened before for some people:
https://github.com/envoyproxy/envoy/blob/83f9eae24f33005bc979e205e8ccdeb5a49ade54/bazel/external/BUILD#L11

Any idea what can I do to make this work locally?

**Comment by mattklein123** [comment]
Seems fine at a high level. Can you add tests, etc. and we can go from there?

/wait

**Comment by kyessenov** [comment]
Very cool. Did you think of the semantic conversion support (https://opentelemetry.io/docs/reference/specification/compatibility/prometheus_and_openmetrics/)? For example, stripping `_bucket` and `_count` suffixes.

**Comment by ohadvano** [comment]
> Very cool. Did you think of the semantic conversion support (https://opentelemetry.io/docs/reference/specification/compatibility/prometheus_and_openmetrics/)? For example, stripping `_bucket` and `_count` suffixes.

Thanks for referring me to this doc, I haven't got to it yet. I'm curious however what is the motivation for supporting this? As I see it, Envoy is not implementing Prometheus semantics in first place? Also if someone chooses this sink then I would assume they would want the semantics to be with OTLP in first place, so what does Prometheus has to do with it? I guess we could add Prometheus sink dedicated for that protocol..

**Comment by kyessenov** [comment]
@ohadvano You are right that Envoy stats pre-date conventions of prometheus. But looking forward, metrics are likely to be added that follow either prometheus or Otel conventions. For example, Istio added `istio_requests_total` in its custom extension, which is in "prometheus" style, so it'd fail "otel" style when converted verbatim. But if you write "otel" style metrics directly, there's no problem with "otel" sink. Perhaps, we can go the other way and instead assume all new metrics are "otel" style and convert to "prometheus" style in the "prometheus" sink. BTW I don't think this need to be solved now, I was curious to hear your thoughts.

**Comment by ohadvano** [comment]
> For example, Istio added istio_requests_total in its custom extension, which is in "prometheus" style, so it'd fail "otel" style when converted verbatim

What would be the expected behavior in such case if OTEL sink is being used?

**Comment by kyessenov** [comment]
Per Otel semantic conventions: `If the counter has a _total suffix, it MUST be removed.`
The other one I know is: duration must be in seconds, and `_milliseconds` suffix must be stripped and unit converted.

**Comment by ohadvano** [comment]
> Per Otel semantic conventions: `If the counter has a _total suffix, it MUST be removed.` The other one I know is: duration must be in seconds, and `_milliseconds` suffix must be stripped and unit converted.

Yeah but I mean, what should happen in case you "violate" this convention and decide to name your counter with ``_total`` suffix anyways? I'm asking this because currently Envoy has some counter stats with this name, for example: [``downstream_cx_total``](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/stats#listener). Is it expected that the collector will drop these metrics? I couldn't find any other doc that describes this, also couldn't find any other doc that specifies these conventions regardless of Prometheus, so will appreciate if you could refer me to any.

**Comment by kyessenov** [comment]
> > Per Otel semantic conventions: `If the counter has a _total suffix, it MUST be removed.` The other one I know is: duration must be in seconds, and `_milliseconds` suffix must be stripped and unit converted.
> 
> Yeah but I mean, what should happen in case you "violate" this convention and decide to name your counter with `_total` suffix anyways? I'm asking this because currently Envoy has some counter stats with this name, for example: [`downstream_cx_total`](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/stats#listener). Is it expected that the collector will drop these metrics? I couldn't find any other doc that describes this, also couldn't find any other doc that specifies these conventions regardless of Prometheus, so will appreciate if you could refer me to any.

In practice, it's not the end of the world. A variety of config options will be created to handle special cases. The benefit of the convention is to reduce configuration overhead, so if we get that right from the start, it can simplify the overall pipeline.

**Comment by ohadvano** [comment]
@mattklein123 I have added tests and I think this is ready for review, thanks.

File: api/envoy/extensions/stat_sinks/open_telemetry/v3/open_telemetry.proto
Comment (mattklein123) [inline api/envoy/extensions/stat_sinks/open_telemetry/v3/open_telemetry.proto:32]: As these are new protos, can you just remove all the not-implemented hide stuff? They can be added back later when the code is added, if it ever is. Same below.

File: docs/root/configuration/observability/stat_sinks/open_telemetry_stat_sink.rst
Comment (mattklein123) [inline docs/root/configuration/observability/stat_sinks/open_telemetry_stat_sink.rst:7]: Can you link to somewhere which explains more about this format?

File: api/envoy/extensions/stat_sinks/open_telemetry/v3/open_telemetry.proto
Comment (ohadvano) [reply api/envoy/extensions/stat_sinks/open_telemetry/v3/open_telemetry.proto]: done

File: api/envoy/extensions/stat_sinks/open_telemetry/v3/open_telemetry.proto
Comment (mattklein123) [inline api/envoy/extensions/stat_sinks/open_telemetry/v3/open_telemetry.proto:37]: If the default value is false these can just be normal bool

File: docs/root/configuration/observability/stat_sinks/open_telemetry_stat_sink.rst
Comment (mattklein123) [inline docs/root/configuration/observability/stat_sinks/open_telemetry_stat_sink.rst:7]: IIRC this is not sphinx external URL format but I can't remember. Please check the rendered docs.

File: source/extensions/stat_sinks/open_telemetry/open_telemetry_impl.cc
Comment (mattklein123) [inline source/extensions/stat_sinks/open_telemetry/open_telemetry_impl.cc:40]: This could cause significant log spam. Either drop this to debug or use a backoff log. Optimally this kind of thind would be covered by a stat.

File: source/extensions/stat_sinks/open_telemetry/open_telemetry_impl.cc
Comment (mattklein123) [inline source/extensions/stat_sinks/open_telemetry/open_telemetry_impl.cc:47]: Same

**Review by mattklein123 — CHANGES_REQUESTED** [review]
LGTM at a high level. Some small comments, thanks.

/wait

File: source/extensions/stat_sinks/open_telemetry/open_telemetry_impl.cc
Comment (ohadvano) [reply source/extensions/stat_sinks/open_telemetry/open_telemetry_impl.cc:29]: Practically it can't, added it as a guard for future changed. Do you suggest to delete this?

Comment (mattklein123) [reply source/extensions/stat_sinks/open_telemetry/open_telemetry_impl.cc:29]: Yes, in general we don't do defensive coding like this.

File: source/extensions/stat_sinks/open_telemetry/open_telemetry_impl.cc
Comment (ohadvano) [reply source/extensions/stat_sinks/open_telemetry/open_telemetry_impl.cc]: done

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Comment by ohadvano** [comment]
@mattklein123 I think this won't merge because of the API approval check, can you please approve it as well?

**Pull request merged**
